### PR TITLE
Refactoring createAppVersion to make it fit into the store

### PR DESCRIPTION
**Pull request by marccampbell**

### CODE REVIEW

File: kotsadm/pkg/store/ocistore/license_store.go [hunk 1]
```diff
@@ -28,5 +28,14 @@ func (s OCIStore) GetLatestLicenseForApp(appID string) (*kotsv1beta1.License, er
...
 
 func (s OCIStore) GetLicenseForAppVersion(appID string, sequence int64) (*kotsv1beta1.License, error) {
-	return nil, ErrNotImplemented
+	appVersion, err := s.GetAppVersion(appID, sequence)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get app version")
+	}
+
+	if appVersion == nil {
+		return nil, nil
```
Comment (divolgin) [inline kotsadm/pkg/store/ocistore/license_store.go:37]: the code that calls this function that don't check if license is nil

File: kotsadm/pkg/store/ocistore/version_store.go [hunk 2]
```diff
@@ -37,6 +37,44 @@ func (s OCIStore) appVersionConfigMapNameForApp(appID string) (string, error) {
...
 
+func (s OCIStore) getLatestAppVersion(appID string) (*versiontypes.AppVersion, error) {
+	configMapName, err := s.appVersionConfigMapNameForApp(appID)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get appversion config map name")
+	}
+
+	configMap, err := s.getConfigmap(configMapName)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get app version config map")
```
Comment (divolgin) [inline kotsadm/pkg/store/ocistore/version_store.go:48]: Should this be `ErrNotFound`?

File: kotsadm/pkg/store/ocistore/version_store.go [hunk 3]
```diff
@@ -270,6 +294,89 @@ func (s OCIStore) GetAppVersionArchive(appID string, sequence int64) (string, er
 	return tmpDir, nil
 }
 
+func (s OCIStore) CreateAppVersion(appID string, currentSequence *int64, appName string, appIcon string, kotsKinds *kotsutil.KotsKinds) (int64, error) {
+	// NOTE that this experimental store doesn't have a tx and it's possible that this
+	// could overwrite if there are multiple updates happening concurrently
+	latestAppVersion, err := s.getLatestAppVersion(appID)
+	if err != nil && err != ErrNotFound {
```
Comment (divolgin) [inline kotsadm/pkg/store/ocistore/version_store.go:301]: `!s.IsNotFound(err)`

File: kotsadm/pkg/handlers/config.go [hunk 4]
```diff
@@ -447,49 +447,52 @@ func decrypt(input string, cipher *crypto.AESCipher) (string, error) {
...
-	keys := []int{}
-	for key, _ := range laterVersions {
+	// ensure that the returned versions array is sorted
+	keys := []string{}
+	for key := range laterVersions {
 		keys = append(keys, key)
 	}
-	sort.Ints(keys)
+
+	// TODO sort by something in kotsutil that i need to write (these are either ints or semvers)
```
Comment (laverya) [inline kotsadm/pkg/handlers/config.go:489]: maybe leftpad with 0s out to 10 or so?

so `0000000009` < `0000000010`

File: kotsadm/pkg/store/ocistore/version_store.go [hunk 5]
```diff
@@ -270,6 +294,89 @@ func (s OCIStore) GetAppVersionArchive(appID string, sequence int64) (string, er
...
+		return nil, errors.New("failed to get configmap name for app version")
+	}
+
+	configMap, err := s.getConfigmap(configMapName)
+	if err != nil {
+		return nil, errors.New("failed to get app version config map")
+	}
+
+	if configMap.Data == nil {
+		return nil, nil // Matching S3PG store
```
Comment (divolgin) [inline kotsadm/pkg/store/ocistore/version_store.go:360]: Callers don't check for nil

File: kotsadm/pkg/store/ocistore/version_store.go [hunk 6]
```diff
@@ -270,6 +294,89 @@ func (s OCIStore) GetAppVersionArchive(appID string, sequence int64) (string, er
...
+	}
+
+	data, ok := configMap.Data[strconv.FormatInt(sequence, 10)]
+	if !ok {
+		return nil, nil // Matching S3PG store
```
Comment (divolgin) [inline kotsadm/pkg/store/ocistore/version_store.go:365]: Callers don't check for nil

File: kotsadm/pkg/handlers/config.go [hunk 7]
```diff
@@ -447,49 +447,52 @@ func decrypt(input string, cipher *crypto.AESCipher) (string, error) {
...
+	if err != nil {
+		return -1, nil, errors.Wrap(err, "failed to get later app versions")
 	}
-	for _, version := range versions {
-		if version.UpdateCursor == thisUpdateCursor && version.Sequence > latestSequenceWithUpdateCursor {
-			latestSequenceWithUpdateCursor = version.Sequence
+
+	// latestSequenceWithThisUpdateCursor is the newest local version
+	// of the same upstream version
+	latestSequenceWithThisUpdateCursor := thisAppVersion.Sequence
```
Comment (divolgin) [inline kotsadm/pkg/handlers/config.go:462]: `thisAppVersion` can be nil

File: kotsadm/pkg/store/ocistore/version_store.go (hunk 2 above)
Comment (marccampbell) [reply kotsadm/pkg/store/ocistore/version_store.go:48]: no, the config map will be created in the get request, so this is an error condition

File: kotsadm/pkg/handlers/config.go (hunk 4 above)
Comment (marccampbell) [reply kotsadm/pkg/handlers/config.go:489]: i'm thinking that we just sort them as semvers?

File: kotsadm/pkg/handlers/config.go (hunk 7 above)
Comment (marccampbell) [reply kotsadm/pkg/handlers/config.go:462]: changed so that `thisAppVersion` cannot be nil

File: kotsadm/pkg/store/ocistore/version_store.go [hunk 8]
```diff
@@ -298,7 +298,7 @@ func (s OCIStore) CreateAppVersion(appID string, currentSequence *int64, appName
 	// NOTE that this experimental store doesn't have a tx and it's possible that this
 	// could overwrite if there are multiple updates happening concurrently
 	latestAppVersion, err := s.getLatestAppVersion(appID)
-	if err != nil && err != ErrNotFound {
+	if !s.IsNotFound(err) {
```
Comment (divolgin) [inline kotsadm/pkg/store/ocistore/version_store.go:301]: still needs `err != nil` :)

Comment (marccampbell) [reply kotsadm/pkg/store/ocistore/version_store.go:301]: @divolgin I think this implementation of `IsNotFound` checks for `nil` in both stores, and that's a good practice.

**Pull request merged**